rfcs/0003-gene-sets/api.md

```diff
@@ -0,0 +1,400 @@
+# Gene Sets
+
+## API Format
+
+**Authors:** [Madison Dunitz](mailto:madison.dunitz@chanzuckerberg.com)
+
+**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Signe Chambers](mailto:schambers@chanzuckerberg.com), [Friendly Tech Person2](mailto:colin.megill@chanzuckerberg.com), [Girish Patangay](mailto:girish.patangay@chanzuckerberg.com), [Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)
+
+## tl;dr
+
+This RFC contains the details about the API for the gene set feature. It will support the export of differential
+expression results and allow a user to upload their own gene sets for analysis.
+
+## Glossery
+
+- Geneset - List of genes
+- Differential expression - A statistic (typically logfold and p-value) describing the difference in gene expression between two sets of cells. It may refer to stats specific to one gene (differential expression value) or many (differential expression analysis). When referring to many genes it is typically capped (eg top 10 most differentially expressed genes)
+- LogFold - A ratio of the difference in expression levels. Typically calculated by taking the log2 of the gene expression value for each cell and then finding the mean of that value for each set of cells. Because the data is already log2 transformed you can then take the difference between the two means as the logfold value
+
+## Problem Statement | Background
+
+Gene sets are the expected output of differential expression analysis in cellxgene. Two (non-overlaping) sets of cells
+are selected and the difference in the expression level for each gene is computed by comparing the average
+expression level for each set of cells. Here I will define the structure of the API to ensure communication between the
+client (where the user is working directly with the data or uploading their own gene sets) and the backend (where data
+is stored to allow it to persist between sessions) is consistent and clear.
+
+## Product Requirements
+
+### Users should be able to export differential expression results
+
+Currently cellxgene users can choose two sets of cells and compute/display the log fold and adjusted p-value of the top ten differentially expressed genes. Additionally we display a histogram of the gene expression levels of the selected cells for each gene in the right side bar.
+
+- [US1] A user should be able to export a gene set, if the gene set was created by differential expression, the pvalue, log fold change, and category label fields will be populated in addition to the gene name. This will be available in a csv file (see data formats for more detail).
+
+### Users should be able to upload their own gene sets and see them in the right side bar with the option to apply them within current cellxgene functionalities, if the same gene ontology was used when generating the current dataset and the gene set uploaded by the user (ie the gene names match)
```

Comment:
```suggestion
### Users can upload their own gene sets and see them in the right side bar with the full range of current cellxgene functionalities.
```

Notes:
- Use "can" to be more assertive about what functionality is available instead "should be able to".
- There is no "option" to apply right? Gene sets will have full functionality in cellxgene explorer and in fact, if the gene set names don't match, you throw an error (later down in the requirements).
- Drop validity requirements for simplicity -- they are written down later as a sub-requirement anyways.